crates/bevy_picking/src/mesh_picking/ray_cast/mod.rs

```diff
@@ -0,0 +1,306 @@
+        self.culled_list.sort_by_key(|(aabb_near, _)| *aabb_near);
+
+        drop(ray_cull_guard);
+
+        // Perform ray casts against the culled entities.
+        let mut nearest_blocking_hit = FloatOrd(f32::INFINITY);
+        let ray_cast_guard = debug_span!("ray_cast");
+        self.culled_list
+            .iter()
+            .filter(|(_, entity)| (settings.filter)(*entity))
+            .for_each(|(aabb_near, entity)| {
+                // Get the mesh components and transform.
+                let Ok((mesh2d, mesh3d, simplified_mesh, has_backfaces, transform)) =
+                    self.mesh_query.get(*entity)
+                else {
+                    return;
+                };
+
+                // Get the underlying mesh handle. One of these will always be `Some` because of the query filters.
+                let Some(mesh_handle) = simplified_mesh
+                    .map(|m| &m.0)
+                    .or(mesh3d.map(|m| &m.0).or(mesh2d.map(|m| &m.0)))
+                else {
+                    return;
+                };
+
+                // Is it even possible the mesh could be closer than the current best?
+                if *aabb_near > nearest_blocking_hit {
+                    return;
+                }
+
+                // Does the mesh handle resolve?
+                let Some(mesh) = self.meshes.get(mesh_handle) else {
+                    return;
+                };
+
+                let backfaces = match has_backfaces {
+                    true => Backfaces::Include,
+                    false => Backfaces::Cull,
+                };
+
+                // Perform the actual ray cast.
+                let _ray_cast_guard = ray_cast_guard.enter();
+                let transform = transform.compute_matrix();
+                let intersection = ray_intersection_over_mesh(mesh, &transform, ray, backfaces);
+
+                if let Some(intersection) = intersection {
+                    let distance = FloatOrd(intersection.distance);
+                    if (settings.early_exit_test)(*entity) && distance < nearest_blocking_hit {
+                        // The reason we don't just return here is because right now we are
+                        // going through the AABBs in order, but that doesn't mean that an
+                        // AABB that starts further away can't end up with a closer hit than
+                        // an AABB that starts closer. We need to keep checking AABBs that
+                        // could possibly contain a nearer hit.
+                        nearest_blocking_hit = distance.min(nearest_blocking_hit);
+                    }
+                    self.hits.push((distance, (*entity, intersection)));
+                };
+            });
+
+        self.hits.retain(|(dist, _)| *dist <= nearest_blocking_hit);
+        self.hits.sort_by_key(|(k, _)| *k);
+        let hits = self.hits.iter().map(|(_, (e, i))| (*e, i.to_owned()));
+        *self.output = hits.collect();
```

Comment:
This should be `self.output.extend(hits)` https://doc.rust-lang.org/std/iter/trait.Extend.html, assuming you're clearing and keeping around the output Vec to avoid allocations.

Currently it's clearing `self.output` for no reason, since it's a new allocation every time from `hits.collect()`

I like the caching the Vec though to avoid allocations :)